2021/day15/day15.java

```diff
@@ -0,0 +1,116 @@
+
+public class day15 {
+    public static class Point {
+        int x;
+        int y;
+
+        Point(int x, int y) {
+            this.x = x;
+            this.y = y;
+        }
+    }
+
+    public static class Position extends Point implements Comparable<Position> {
+        int risk;
+
+        Position(int x, int y, int risk) {
+            super(x, y);
+            this.risk = risk;
+        }
+
+        @Override
+        public int compareTo(Position other) {
+            return Integer.compare(this.risk, other.risk);
+        }
+    }
+
+    public static ArrayList<ArrayList<Integer>> readInput(String filename) {
+        ArrayList<ArrayList<Integer>> mapOfRisks = new ArrayList<>();
+        Scanner sc;
+        try {
+            sc = new Scanner(new File(filename));
+            while (sc.hasNextLine()) {
+                mapOfRisks.add(splitRow(sc.nextLine()));
+            }
+            sc.close();
+        } catch (Exception e) {
+            System.out.println("Error handling input " + e);
+        }
+        return mapOfRisks;
+    }
+
+    public static ArrayList<Integer> splitRow(String row) {
+        ArrayList<Integer> rowOfRisks = new ArrayList<>();
+        for (String s : new ArrayList<String>(Arrays.asList(row.split("")))) {
+            rowOfRisks.add(Integer.parseInt(s));
+        }
+        return rowOfRisks;
+    }
+
+    public static int getRisk(ArrayList<ArrayList<Integer>> mapOfRisks, int x, int y) {
+        int tmpRisk = mapOfRisks.get(y % mapOfRisks.size()).get(x % mapOfRisks.get(0).size())
+                + (x / mapOfRisks.get(0).size()) + (y / mapOfRisks.size());
+        return tmpRisk < 10 ? tmpRisk : ((tmpRisk % 10) + 1);
+    }
+
+    public static ArrayList<Point> getNeighbours(int x, int y, int width, int height) {
+        ArrayList<Point> neighbours = new ArrayList<>();
+        if (x > 0) {
+            neighbours.add(new Point(x - 1, y));
+        }
+        if (x < width - 1) {
+            neighbours.add(new Point(x + 1, y));
+        }
+        if (y > 0) {
+            neighbours.add(new Point(x, y - 1));
+        }
+        if (y < height - 1) {
+            neighbours.add(new Point(x, y + 1));
+        }
+        return neighbours;
+    }
+
+    public static int calculateRisk(boolean isPart2) {
+        ArrayList<ArrayList<Integer>> mapOfRisks = readInput("data.txt");
+        PriorityQueue<Position> positionQueue = new PriorityQueue<>();
+        int height = mapOfRisks.size();
+        int width = mapOfRisks.get(0).size();
+        if (isPart2) {
+            height *= 5;
+            width *= 5;
+        }
+        ArrayList<ArrayList<Boolean>> isVisited = new ArrayList<>();
+        for (int i = 0; i < height; i++) {
+            ArrayList<Boolean> line = new ArrayList<>(Collections.nCopies(width, false));
+            isVisited.add(line);
+        }
+        int x = 0;
+        int y = 0;
+        int risk = 0;
+        while (x != width - 1 || y != height - 1) {
```

Comment:
If you wanted to be as efficient as possible, you would return the risk inside the `for` loop whenever you see the bottom-right corner being a neighbour. Otherwise you may be going through a lot of unnecessary steps in the meantime. I don't deem that a necessary change though, up to you.

Also, it's worth noting that my approach would not work for all types of shortest-path problems. Do you know why it works here?

Reply:
No idea :(

Reply:
The key is that in this problem you consider the risk levels to be the "costs of edges" of your graph and thus each "edge" going into the bottom-right corner has the same cost. This means that if you are at a neighbour of your goal vertex, it doesn't matter which neighbour it is, the one remaining "edge" will have the same cost and thus you can return the result immediately. Contrast this to the following case (where A is the source and D is the goal):

```
A ----3----> B
|                      |
|                      |
10                 30
|                      |
v                    v
C ----1----> D
```

Here, you would consider vertex `B` before `C` in your priority queue, but even though it's the neighbour of D, the cost of the path A->B->D is higher than A->C->D.